packages/dds/tree/src/simple-tree/api/schemaFactory.ts

```diff
@@ -277,6 +278,16 @@ export class SchemaFactory<
 	 */
 	public readonly string = stringSchema;
 
+	public limitedString<TString extends string>(): TreeNodeSchemaNonClass<
```

Comment:
This schema always has the same identifier, so if you make actual new schema objects to return here (which I believe this does) it makes it a hard error to use two kinds of strings in a single document.

If you used a different identifier for each one, then it would break because we would have no way to get from the instance back to the type, which we require in several places (like Tree.schema).

The only way to do this that won't break all other strings would be to return the same string schema object by type cast. While I suspect that would mostly work, it would still break things, like if you tried to narrow a union of limited strings using Tree.is, it would look like it would work at compile time, but would always return true at runtime.

Additionally our schema system guarantees data stays in schema. It needs to be able to enforce this with runtime information not just compile time type annotations. Otherwise other clients, or past or future version of the app (or application bugs), or operations like moves between different locations in the tree could produce out of schema documents.

Given that there is no possible way we can guarantee this type narrowing holds, I don't think it belongs in the same kind of API as the rest of our schema system where we actually do runtime enforcement, compatibility checking on document open, offer schema evolution and schema export tools etc.


If you want actually enforced schema for something like this, consider using the approach [enumFromStrings](https://fluidframework.com/docs/api/v2/tree#enumfromstrings-function) uses. https://fluidframework.com/docs/api/v2/tree#adaptenum-function might also be useful.


If you really do want to have the actual tree content store a string, and have no validation of the string against your type constraints (just rely on the application code to maintain that invariant, including across moves, version and schema changes, merges etc), I'm working on an API to do things like that. Generally though APIs like that will need a clear disclaimer, and be separated from the core schema APIs which actually have validation and enforcement.